# My first DQ project (on most used free apps)

**with this project I will try to gain a view on which free apps are most sought after in order to monetize the possibility to introduce adds on a free app we are trying to develop**

- opening the Apple Dataset [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home):

In [1]:
from csv import reader
opened_file_A = open('AppleStore.csv', encoding="utf8")
read_file_A = reader(opened_file_A)
Apps_Ds = list(read_file_A)

- opening the Google Play Dataset [link](https://www.kaggle.com/lava18/google-play-store-apps/home):

In [2]:
opened_file_G =  open('googleplaystore.csv', encoding="utf8")
read_file_G = reader(opened_file_G)
Google_Ds = list(read_file_G)


- creating a function to xplore the data:

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') #adds a new (empty) line
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [4]:
print(Apps_Ds[0]) #printing the header of the Apple Dataset

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [5]:
print(Google_Ds[0]) #printing the header of the Google Dataset

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


**Now we will, before the analysis, have to do some data cleaning to our 2 Datasets. And remove duplicates or typos in the info, or info that really doesnt matter, like apps in other language other than English!**

In [6]:
print(Google_Ds[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


- the Google Dataset has an error in row 10473. Were the Category of the app doesnt appear, instead in its place we have the rating. We will have to delete this row:

In [7]:
del Google_Ds[10473]

- Scraping for Duplicate entries in our Dataset. Because As read in the Kaggle Discussion area, for the Google Play store Dataset, this Dataset as a number os duplicate entries on it. 
- As example I will show some duplicates for the Instagram app:

In [8]:
for column in Google_Ds:
    header = Google_Ds[0]
    name = column[0]
    if name == 'Instagram':
        print(header)
        print(column)
        print("\n")

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Cate

- Now we should and ought to count how many duplicate entries we have in this Dataset of Google Play Store. By Looping trough the entire Dataset and verifying if the variable *name*, that should correspond to the name of an app is already in the empty list we created for the unique_apps, and if it isn't to be part of that list, and once that very same app is encountered again, and is part of the unique_apps list, should then be part of the duplicate_apps list. We then simply return the total number of those duplicate entries, and also print some examples of them:

In [9]:
duplicate_apps = []
unique_apps = []

for column in Google_Ds:
    name = column[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Nº of Duplicate Apps:', len(duplicate_apps))
print('\n')
print('Some examples of Duplicate Entries:', duplicate_apps[:10])

Nº of Duplicate Apps: 1181


Some examples of Duplicate Entries: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


- This Duplicate Entries won't be, obsviously, randomly deleted! I will use the info that's differentiating them from one another, the column *Reviews*, and assume that the ones that have the more *Reviews* are the ones that are more updated. Being therefore the ones that should stay in our Dataset and all the other duplicates deleted! 

- We will start by first sorting and saving the apps that have the highst number of reviews.
- First we will creat an empty dict with the name of *reviews_max*, where we will save the name of the app as a key, and the total number of reviews as the value of that unique key
- Then we will loop through the Google_Ds and isolate the variable *name* which corresponds to the name of the app, and the variable *n_reviews* which corresponds to the nº of the reviews a certain app has.

    - Using an if Statement we will assess if the app already exists in the dict we created and if the key:value is inferior to the the biggest review it will have the biggest review as the new value for the key name app. If the app is not in the reviews_max dict it will had it to the dict under the form key(name):value(n_reviews)

In [10]:
reviews_max  = {}

for column in Google_Ds[1:]:
    name = column[0]
    n_reviews = float(column[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews


#print(reviews_max)

print(len(reviews_max))
print(len(Google_Ds[1:]) - 1181)

9659
9659


- We can check above that now we have the same number of unique apps in our *reviews_max* dict as we should have in our Google_Ds: 9659

## Now lets incorporate our unique entries into our Google_Ds and delete the Duplicates:
- first lets creat 2 empty lists and assign them to the variables *Google_Ds_clean* and *already_added*. The first one will store our newly cleaned and unique rows, and the second will save the name of the apps we add to the list *Google_Ds_clean*

- Afterwards we will loop through our *Google_Ds* and isolate the variables *name* and *n_reviews*
    - For each iteration we will append the rows/columns to our *Google_Ds_clean* list, if the variable *n_reviews* is equal to the value we had in our dict *reviews_max*, that should have the **key**(name of the app)**:value**(max review of that very app. And we also have to check if that app isnt already in the *already_added* list! We have to validate this condition because we have apps that have duplicate entries and that cannot be sorte through the nº of their reviews, because they simple have the same nº of reviews for all the duplicated entries! So one way to not include the duplicate entries is by when in each iteration we had that app in our *Google_Ds_clean* we also include its name in the list *already_added* and once its in this list even if it validates the first condition: **if n_reviews == reviews_max[name]** it will not be included because the name is already in the *already_added* list, therefore solving this issue!

In [11]:
Google_Ds_clean = []
already_added = []

for column in Google_Ds[1:]:
    name = column[0]
    n_reviews = float(column[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        Google_Ds_clean.append(column)
        already_added.append(name)

print(Google_Ds_clean[:10])
print('\n')
print(len(Google_Ds_clean))

explore_data(Google_Ds_clean, 0, 3, True) 
    

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up'], ['Smoke Effect Photo Maker - Smoke Editor', 'ART_AND_DESIGN', '3.8', '178', '19M', '50,000+', '

#### Perfect! We can validate that our new and cleaned *Google_Ds_clean* has all the rows it should have: 9659

### Scouring through the Data of both Datasets one can check that there are several apps that are for non-english users!


In [12]:
print(Google_Ds_clean[4412][0])
print(Google_Ds_clean[7940][0])
print('\n')
print(Apps_Ds[814][1])

中国語 AQリスニング
لعبة تقدر تربح DZ


爱奇艺PPS -《欢乐颂2》电视剧热播


  - Bellow we will try to creat a function that uses the ASCII counting system (0-127) to count each character of an inputed string and sort which ones bellong to the English common set of characters, and which ones dont:

In [13]:
def check_str_Eng(a_string):
    
    for character in a_string:
        if ord(character) > 127:
            return False
    else: 
        return True
    '''esta linha do else tem que estar fora da 1º condição if. Isto vai fzr 1
           validação da 1º condição em cada loop e, se encontra dentro d cada
           iteração, seja ela a 1º iteração p ex, 1 caracter q seja verdadeiro 
           ele pára a iteração e da a string como verdadeira e correndo os 
           outros caracteres da string. Dai que temos que os correr todos 1º,
           para avaliar se não pertencem, e só dp de os corrermos tds é que 
           podemos validar a string como True, sim, pertence ao sistema ASCII'''
        

In [14]:
print(check_str_Eng('Instagram'))

True


In [15]:
print(check_str_Eng('爱奇艺PPS -《欢乐颂2》电视剧热播'))

False


In [16]:
print(check_str_Eng('Docs To Go™ Free Office Suite'))

False


In [17]:
print(check_str_Eng('Instachat 😜'))

False


### Some of the *"English_Apps"*  are being tagged as not bellonging to the Dataset of English language market target. This happens due to special characters embedded on their names, such as ™ or emojis
    - We will have to create a different function, one that takes into account this special characters, and we will do so by first defining a cutoff. 
    - Our new function should accept up to a max of 3 special characters per string in order to catalogue it as an *"English_Apps"*. Anything more than this and the app should fall into the category of non-english!

In [18]:
def check_str_Eng_1(a_string):
    n_Falses = 0
    for character in a_string:
        if ord(character) > 127:
            n_Falses += 1
            if n_Falses > 3:
                return False
    else:
        return True

In [19]:
print(check_str_Eng_1('Docs To Go™ Free Office Suite'))

True


In [20]:
print(check_str_Eng_1('Instachat 😜'))

True


In [21]:
print(check_str_Eng_1('爱奇艺PPS -《欢乐颂2》电视剧热播'))

False


## PERFECT, VALIDATED! 

- Applying our function to both Datasets:

In [22]:
Google_Ds_clean_1 = []
Apps_Ds_clean_1 = []


for app in Google_Ds_clean:
        name = app[0]
        if check_str_Eng_1(name):
            Google_Ds_clean_1.append(app)
            
for app in Apps_Ds:
    name = app[1]
    if check_str_Eng_1(name):
        Apps_Ds_clean_1.append(app)

explore_data(Google_Ds_clean_1, 0, 4, True)
print('\n')
print('\n')
explore_data(Apps_Ds_clean_1, 0, 4, True)

        

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 9614
Number of columns: 13




['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'la

### Now we are moving to our last stage of the data cleaning process! We already:
    - Removed inncurate data;
    - Removed duplicate app entries;
    - Removed non-English apps.
### Now its time to remove all the apps from both Datasets that arent for free:

In [23]:
Google_Ds_clean_2 = []
Apps_Ds_clean_2 = []


for column in Google_Ds_clean_1[1:]:
    type_price = column[6]
    price = column[7]
    if type_price == 'Free' and price == '0':
        Google_Ds_clean_2.append(column)
        
for column in Apps_Ds_clean_1[1:]:
    price = column[4]
    if price == '0.0':
        Apps_Ds_clean_2.append(column)

explore_data(Google_Ds_clean_2, 0, 4, True)
print('\n')
print('\n')
explore_data(Apps_Ds_clean_2, 0, 4, True)
       

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 8862
Number of columns: 13




['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '

#### We were left with **3222 rows in the Apps_Ds_clean_2** and **8862 in the Google_Ds_clean_2**

## In this phase we will need to evaluate which of the free apps from both Datasets are most common their own markets! We will do so by first constructing 2 frequency tables for each of the 2 Datasets and their respective genres 

In [24]:
frequency_table_Google_Ds = {}
frequency_table_Apps_Ds = {}

for column in Google_Ds_clean_2[1:]:
    app_genre = column[1]
    
    if app_genre in frequency_table_Google_Ds:
        frequency_table_Google_Ds[app_genre] += 1
    else:
        frequency_table_Google_Ds[app_genre] = 1
     
    
for column in Apps_Ds_clean_2[1:]:
    app_genre = column[11]
    
    if app_genre in frequency_table_Apps_Ds:
        frequency_table_Apps_Ds[app_genre] += 1   
    else:
        frequency_table_Apps_Ds[app_genre] = 1

print('Google:',  frequency_table_Google_Ds)
print('\n')
print('Apps:', frequency_table_Apps_Ds)



Google: {'HEALTH_AND_FITNESS': 273, 'BUSINESS': 407, 'BEAUTY': 53, 'WEATHER': 71, 'MEDICAL': 313, 'VIDEO_PLAYERS': 159, 'GAME': 862, 'LIBRARIES_AND_DEMO': 83, 'HOUSE_AND_HOME': 73, 'NEWS_AND_MAGAZINES': 248, 'EVENTS': 63, 'PARENTING': 58, 'ENTERTAINMENT': 85, 'EDUCATION': 103, 'BOOKS_AND_REFERENCE': 190, 'SOCIAL': 236, 'COMMUNICATION': 287, 'SHOPPING': 199, 'TRAVEL_AND_LOCAL': 207, 'LIFESTYLE': 346, 'MAPS_AND_NAVIGATION': 124, 'PHOTOGRAPHY': 261, 'FINANCE': 328, 'AUTO_AND_VEHICLES': 82, 'DATING': 165, 'ART_AND_DESIGN': 55, 'SPORTS': 301, 'COMICS': 55, 'PERSONALIZATION': 294, 'PRODUCTIVITY': 345, 'FAMILY': 1675, 'TOOLS': 750, 'FOOD_AND_DRINK': 110}


Apps: {'Music': 66, 'Utilities': 81, 'Social Networking': 105, 'Sports': 69, 'Games': 1874, 'Health & Fitness': 65, 'Travel': 40, 'Finance': 36, 'Book': 14, 'Reference': 18, 'Lifestyle': 51, 'Navigation': 6, 'Photo & Video': 160, 'Shopping': 84, 'Medical': 6, 'News': 43, 'Education': 118, 'Productivity': 56, 'Business': 17, 'Entertainment':

- ##### Now that we know that one of the most common free apps both in Apps and Google are of the genre *Games*:

    - Our goal will be to develop one *Game* app and provide it in the *Google App Store*. 
    - We will do so by undertaken the minimun resources as this should be a sample pilot test that, if results in a good response from the market ougth to be further developed.
        - If it reaches profitability in 6 months we should direct it to the *App Store* market as well! 

- We can now creat a generic function, that we can use to extract a frequency table, expressed in percentages, out of any Dataset and of any feature of it! 

- That's what we will try to do bellow:

In [25]:
def freq_table(dataset, index):
    feature_percentages = {}
    n_apps = len(dataset)
    features = {}
    for row in dataset:
        feature = row[index]
        if feature not in features:
            features[feature] = 1
        else:
            features[feature] += 1
      
    for key in features:
        value = features[key]
        feature_percentages[key] = 100 * value/n_apps
    
    return feature_percentages  

   - Above we created a function that accpets 2 inputs (a dataset and an index. 
    - Created 2 empty dicts (feature_percentages + features)
    - Looped through each row of our choosen dataset, and for each loop we assigned our index row choosen to a variable named feature
        - In each loop we check if the feature we've choosen is not in the empty dict we created named features, if it isn't we need to assigned that feature there as a *Key*. If in the next loop it finds there that feature as a *Key* it should add a value to it
        
    - After this we created another for Loop to iterate through the new and fully filled *features* dict. 
        - for each iteration our loop will extract the *Value* out of each *Key* and asign it to a variable named **value**
        - and also for each iteration it will fill our other empty dict **feature_percentages** with each *Key* and it corresponding *Value* but express as the percentage that it represents in the full length of our Dataset

- Now let's see if it's working:

In [26]:
freq_table(Google_Ds_clean_2, 1)

{'ART_AND_DESIGN': 0.631911532385466,
 'AUTO_AND_VEHICLES': 0.9252990295644324,
 'BEAUTY': 0.598059128864816,
 'BOOKS_AND_REFERENCE': 2.143985556307831,
 'BUSINESS': 4.592642744301512,
 'COMICS': 0.6206273978785827,
 'COMMUNICATION': 3.2385466034755135,
 'DATING': 1.8618821936357481,
 'EDUCATION': 1.162265854208982,
 'ENTERTAINMENT': 0.9591514330850823,
 'EVENTS': 0.7109004739336493,
 'FAMILY': 18.900925299029563,
 'FINANCE': 3.7011961182577298,
 'FOOD_AND_DRINK': 1.2412547957571654,
 'GAME': 9.726923944933423,
 'HEALTH_AND_FITNESS': 3.080568720379147,
 'HOUSE_AND_HOME': 0.8237418190024826,
 'LIBRARIES_AND_DEMO': 0.9365831640713157,
 'LIFESTYLE': 3.9043105393816293,
 'MAPS_AND_NAVIGATION': 1.3992326788535319,
 'MEDICAL': 3.53193410065448,
 'NEWS_AND_MAGAZINES': 2.7984653577070637,
 'PARENTING': 0.6544798013992327,
 'PERSONALIZATION': 3.3175355450236967,
 'PHOTOGRAPHY': 2.945159106296547,
 'PRODUCTIVITY': 3.893026404874746,
 'SHOPPING': 2.245542766869781,
 'SOCIAL': 2.663055743624464,
 

##### PERFECT!

#### Aplying our *freq_table* function to another function (display_table()), in order to see/sort our frequency table in a particular order...:

In [27]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [28]:
print(display_table(Google_Ds_clean_2, 1))
print('\n')
print(display_table(Google_Ds_clean_2, 9))
print('\n')
print(display_table(Apps_Ds_clean_2, 11))

FAMILY : 18.900925299029563
GAME : 9.726923944933423
TOOLS : 8.463100880162491
BUSINESS : 4.592642744301512
LIFESTYLE : 3.9043105393816293
PRODUCTIVITY : 3.893026404874746
FINANCE : 3.7011961182577298
MEDICAL : 3.53193410065448
SPORTS : 3.39652448657188
PERSONALIZATION : 3.3175355450236967
COMMUNICATION : 3.2385466034755135
HEALTH_AND_FITNESS : 3.080568720379147
PHOTOGRAPHY : 2.945159106296547
NEWS_AND_MAGAZINES : 2.7984653577070637
SOCIAL : 2.663055743624464
TRAVEL_AND_LOCAL : 2.3358158429248475
SHOPPING : 2.245542766869781
BOOKS_AND_REFERENCE : 2.143985556307831
DATING : 1.8618821936357481
VIDEO_PLAYERS : 1.7941773865944481
MAPS_AND_NAVIGATION : 1.3992326788535319
FOOD_AND_DRINK : 1.2412547957571654
EDUCATION : 1.162265854208982
ENTERTAINMENT : 0.9591514330850823
LIBRARIES_AND_DEMO : 0.9365831640713157
AUTO_AND_VEHICLES : 0.9252990295644324
HOUSE_AND_HOME : 0.8237418190024826
WEATHER : 0.8011735499887158
EVENTS : 0.7109004739336493
PARENTING : 0.6544798013992327
ART_AND_DESIGN : 0.63

#### Some first impressions on the FT (frequency table) of the DataSet Apps_Ds_clean_2 (Apple App Store_free_english_apps):

   - Analyzing the DataSet one can conclude that the most common genre in this sample is *Games* representing more than half (58%) of all the apps in our DataSet. Followed, far behind, by the genre *Entertainment* with approximately 8% (7.88%);
   - Another interesting pattern in the DataSet is that there are almost no English free apps in the genre *Navigation*! Due to intensive competition from the free Google App *Google Maps* (?). Probably won't be either much market, I would risk almost none, for the paid ones in this genre! 
    *Medical* is also a genre were developers do not offer much of free English Apps, maybe because this is more of a market that first has to "pass through" a middle man (doctor?) or an institution (hospital?) in order to monitize their products. Therefore, making it difficult for this market to profit and to rev up sales targeting directly their consumers, who first are not the most suitable ones to choose on their own this array of products, and second, in a legal matter, couldn't even do it all by themselfs without proper medical advice;
   - Generally we can see that most of the free English apps are designed and developed towards entertainment rather than pratical purposes. With the top 6 being occupied by app genres such as *Games*, *Entertainment*, *Photo & Video*, *Education*, *Social Networking* and *Shoping* and the bottom 6 *Reference*, *Business*, *Book*, *Navigation*, *Medical*, *Catalogs*;
   - One, based only in this FT, could not blindly recommend an app profile for this market. The info, I think, its insufficient to take any final conclusion. We could make a wrong conclusion by saying that if there's a lot of apps developed in the *Games* genre then maybe thats were there are more users or more market share. This could be right, but it could also be profoundly wrong! This could mean, for example, that Game apps are easier to develop, therefore being available for a big array of developers to jump in and flood the market, trying their luck. It could also mean that developing Games is cheaper, making this genre appealing for a pilot app for future development! **To better grasp this matter I think we should cross this info with others (rating_count_tot(?)**


#### Some first impressions on the FT of the DataSet Google_Ds_clean_2 (Google Play Store_free_english_apps):

   - The most common **Genre** of apps here is *Tools* with 8,4% followed by *Entertainment* with aprox 6,07%, and in the third position comes Education counting ca. 5,35% of all free english apps developed; 
   - In the **Category** designation *Tools* has also a prime position, coming in third with approximately 8,5%, in second is *Games* with 9,7% of all developed apps, and in the first position highlighted with almost 19% (18,9%) is the **Category** *Family*;
   - Education and Productivity orientated Apps, play in Google Play Store a major role. With *Tools*, *Family*, and *Education* being in the top 3 of both **Genres** and **Category**;
   - Comparing this patterns with the ones discovered in the **Apps_Ds_clean_2** DataSet we can conclude that although the **Genre** *Games* play an important role in both DataSet's, it is far less important in Google Play Store than in Apple Store (9,7% Vs. 58%). We can also assume that Apple Store is more Entertainment orientated compared to the Google one.
   - Once more, for a better understandment and a more useful recommendation of which app to develop, we should cross this info with other such has *Installs* and *Reviews*. With this 3 sections we should get a pretty accurate view of whats the most popular genre of Apps, or which kind of App should we be targeting our efforts into.  

In [29]:
genres = freq_table(Apps_Ds_clean_2, 11)
print(genres)

{'Music': 2.0484171322160147, 'Utilities': 2.5139664804469275, 'Business': 0.5276225946617008, 'Sports': 2.1415270018621975, 'Games': 58.16263190564867, 'Health & Fitness': 2.017380509000621, 'Navigation': 0.186219739292365, 'Travel': 1.2414649286157666, 'Finance': 1.1173184357541899, 'Book': 0.4345127250155183, 'Reference': 0.5586592178770949, 'Lifestyle': 1.5828677839851024, 'Social Networking': 3.2898820608317814, 'Photo & Video': 4.9658597144630665, 'Shopping': 2.60707635009311, 'Medical': 0.186219739292365, 'News': 1.3345747982619491, 'Education': 3.6623215394165114, 'Productivity': 1.7380509000620732, 'Entertainment': 7.883302296710118, 'Food & Drink': 0.8069522036002483, 'Weather': 0.8690254500310366, 'Catalogs': 0.12414649286157665}


In [30]:
for genre in genres:
    total = 0 #this variable will store the total nº of ratings for each genre
    len_genre = 0 #this variable will store the total nº of apps of each genre
    
    for row in Apps_Ds_clean_2:
        genre_app = row[11]
        if genre_app == genre:
            user_ratings_app = float(row[5])
            total += user_ratings_app
            len_genre += 1
            
    avg_user_ratings = total/len_genre
    
    print(genre, ':', avg_user_ratings)



    

Music : 57326.530303030304
Utilities : 18684.456790123455
Business : 7491.117647058823
Sports : 23008.898550724636
Games : 22788.6696905016
Health & Fitness : 23298.015384615384
Navigation : 86090.33333333333
Travel : 28243.8
Finance : 31467.944444444445
Book : 39758.5
Reference : 74942.11111111111
Lifestyle : 16485.764705882353
Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Shopping : 26919.690476190477
Medical : 612.0
News : 21248.023255813954
Education : 7003.983050847458
Productivity : 21028.410714285714
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Weather : 52279.892857142855
Catalogs : 4004.0


#### After doing this analysis, we can see that the most "popular" **Genre** is the *Navigation*. With an average of 86090 ratings per app developed. This is a very interesting result, since this one of the least developed apps_genre, coming in the top 3 of the last ones, followed by *Medical* and *Catalogs*.
#### But I would no recommend a *Navigation* app, I would instead choose a *Social Networking* orientated one. It comes in 3rd has one of the most rated genres in average (71548) and on the other side is also one of the most popular genres in terms of number of apps developed, rating 5h (3.3%)

In [31]:
Categories = freq_table(Google_Ds_clean_2, 1)
print(Categories)

{'FAMILY': 18.900925299029563, 'HEALTH_AND_FITNESS': 3.080568720379147, 'BUSINESS': 4.592642744301512, 'BEAUTY': 0.598059128864816, 'WEATHER': 0.8011735499887158, 'MEDICAL': 3.53193410065448, 'GAME': 9.726923944933423, 'LIBRARIES_AND_DEMO': 0.9365831640713157, 'HOUSE_AND_HOME': 0.8237418190024826, 'FINANCE': 3.7011961182577298, 'EVENTS': 0.7109004739336493, 'PARENTING': 0.6544798013992327, 'ENTERTAINMENT': 0.9591514330850823, 'EDUCATION': 1.162265854208982, 'TRAVEL_AND_LOCAL': 2.3358158429248475, 'SOCIAL': 2.663055743624464, 'COMMUNICATION': 3.2385466034755135, 'SHOPPING': 2.245542766869781, 'VIDEO_PLAYERS': 1.7941773865944481, 'LIFESTYLE': 3.9043105393816293, 'PHOTOGRAPHY': 2.945159106296547, 'MAPS_AND_NAVIGATION': 1.3992326788535319, 'AUTO_AND_VEHICLES': 0.9252990295644324, 'DATING': 1.8618821936357481, 'ART_AND_DESIGN': 0.631911532385466, 'NEWS_AND_MAGAZINES': 2.7984653577070637, 'COMICS': 0.6206273978785827, 'PERSONALIZATION': 3.3175355450236967, 'PRODUCTIVITY': 3.893026404874746, 

In [32]:
for category in Categories:
    total = 0
    len_category = 0
    
    for row in Google_Ds_clean_2:
        category_app = row[1]
        if category_app == category:
            n_installs = row[5]
            # Remove any + or , character, and then convert the string to a float.
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            n_installs = float(n_installs)
            total += n_installs
            len_category += 1
            
    avg_category_installs = total/len_category
    print(category, ':', (round(avg_category_installs, 2)))

        
            
    

FAMILY : 3697848.17
HEALTH_AND_FITNESS : 4188821.99
BUSINESS : 1712290.15
BEAUTY : 513151.89
WEATHER : 5074486.2
MEDICAL : 120550.62
GAME : 15588015.6
LIBRARIES_AND_DEMO : 638503.73
HOUSE_AND_HOME : 1331540.56
FINANCE : 1387692.48
EVENTS : 253542.22
PARENTING : 542603.62
ENTERTAINMENT : 11640705.88
EDUCATION : 1833495.15
TRAVEL_AND_LOCAL : 13984077.71
SOCIAL : 23253652.13
COMMUNICATION : 38456119.17
SHOPPING : 7036877.31
VIDEO_PLAYERS : 24727872.45
LIFESTYLE : 1437816.27
PHOTOGRAPHY : 17840110.4
MAPS_AND_NAVIGATION : 4056941.77
AUTO_AND_VEHICLES : 647317.82
DATING : 854028.83
ART_AND_DESIGN : 2021626.79
NEWS_AND_MAGAZINES : 9549178.47
COMICS : 817657.27
PERSONALIZATION : 5201482.61
PRODUCTIVITY : 16787331.34
SPORTS : 3638640.14
TOOLS : 10801391.3
BOOKS_AND_REFERENCE : 8767811.89
FOOD_AND_DRINK : 1924897.74


### After these results in both analysis (Apple Store + Google Play Store), I would recommend, as a free english-app our company should develop, an app orientated towards the *Social Network* ecosystem! Since in both DataSet's this genre/category is well positioned as beeing one of them that most buzz creats around users. Positioning itself at number 3 in Google Play Store, as one of the genres/categories with more installs in average per user (23253652.13). And if we take into consideration that the other first two genres/categories (*Communication* and *Video_Players*) may share apps with the *Social* genre and vice-versa, due to their similarity apps may have leaked and leaped from one of these 3 genres to another, we can attest the true value of the genre/category *Social Network/Social*

### I would recommend the development of an app that could be used as a video-recording/Vlog and that could creat a social network around it and its users, sharing and commenting their content. Giving it also the ability to be shared in the more mainstream Social-Networks. Monetizing it in the first stage with adds, and in a second stage with fees, unlocking new abilities like for example UHD recording, following up to 100.000 friends or new recording filters. Taking full advantage of the 3 areas, *Communication*, *Video* and *Social* and also flowing with this new wave of a more narcissistic and influencers hostage society! 